library/ssl_tls13_generic.c

```diff
@@ -136,6 +136,84 @@ void mbedtls_ssl_tls13_add_hs_hdr_to_checksum( mbedtls_ssl_context *ssl,
 }
 
 #if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
+/* mbedtls_ssl_tls13_parse_sig_alg_ext()
+ *
+ * enum {
+ *    ....
+ *   ecdsa_secp256r1_sha256( 0x0403 ),
+ *   ecdsa_secp384r1_sha384( 0x0503 ),
+ *   ecdsa_secp521r1_sha512( 0x0603 ),
+ *    ....
+ * } SignatureScheme;
+ *
+ * struct {
+ *    SignatureScheme supported_signature_algorithms<2..2^16-2>;
+ * } SignatureSchemeList;
+ */
+int mbedtls_ssl_tls13_parse_sig_alg_ext( mbedtls_ssl_context *ssl,
+                                         const unsigned char *buf,
+                                         const unsigned char *end )
+{
+    const unsigned char *p = buf;
+    size_t sig_algs_len = 0;
+    uint16_t *sig_algs;
+    const unsigned char *sig_algs_end;
+    uint16_t sig_alg;
+    uint32_t common_idx = 0; /* iterate through received signature schemes list */
+
+    /* skip 2 bytes of signature algorithms length */
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+    sig_algs_len = MBEDTLS_GET_UINT16_BE( p, 0 );
+    p += 2;
+
+    sig_algs = mbedtls_calloc( MBEDTLS_SIG_ALGS_SIZE, sizeof( uint16_t ) );
+    if( sig_algs == NULL )
+        return( MBEDTLS_ERR_SSL_ALLOC_FAILED );
+
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, sig_algs_len );
+    sig_algs_end = p + sig_algs_len;
+    while( p < sig_algs_end && common_idx + 1 < MBEDTLS_SIG_ALGS_SIZE )
+    {
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, sig_algs_end, 2 );
+        sig_alg = MBEDTLS_GET_UINT16_BE( p, 0 );
+        p += 2;
+
+        MBEDTLS_SSL_DEBUG_MSG( 4, ( "received signature algorithm: 0x%x",
+                                    sig_alg ) );
+
+        if( mbedtls_ssl_sig_alg_is_offered( ssl, sig_alg ) &&
+            mbedtls_ssl_sig_alg_is_supported( ssl, sig_alg ) )
+        {
+            sig_algs[common_idx] = sig_alg;
+            common_idx += 1;
+        }
+    }
+    /* Check that we consumed all the message. */
+    if( p != end )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1,
+            ( "Signature algorithms extension length misaligned" ) );
+        MBEDTLS_SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_DECODE_ERROR,
+                                      MBEDTLS_ERR_SSL_DECODE_ERROR );
+        mbedtls_free( sig_algs );
+        return( MBEDTLS_ERR_SSL_DECODE_ERROR );
+    }
+
+    if( common_idx == 0 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "no signature algorithm in common" ) );
+        MBEDTLS_SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE,
+                                      MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
+        mbedtls_free( sig_algs );
+        return( MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
+    }
+
+    sig_algs[common_idx] = MBEDTLS_TLS1_3_SIG_NONE;
+    ssl->handshake->sig_algs = sig_algs;
```

Comment:
```suggestion
    ssl->handshake->received_signature_schemes_list = sig_algs;
```

Reply:
I do not also think that it is a good idea to use the `sig_algs` field of the handshake context for two different things. Its purpose is to store the configured list of signature algorithms not the list of signature algorithms we received from the peer. For the name I'd prefer `received_sig_algs` as "signature scheme" refer to a type in the spec not to values.

Reply:
Agree.